fix time-sensitive sensor derivation when rebinding to old sensors

ahgittin opened the pull request:
before this, autoscaler might scale out on rebind, because the
last total reqs might be 100, if data is v old, current value might be 10000;
a recent timestamp is attached to initial value, meaning it computes 9900 new reqs in a 1s window.

this forces an invalid timestamp for enrichers

better fix would be to store timestamp on sensors themselves

ahgittin commented:
took a while to get all the tests happy with this change but finally there

ahgittin commented:
@aledsage @grkvlt can you look at this PR also?  without it, odd things happen on rebind with policies.

aledsage commented on policy/src/main/java/brooklyn/enricher/RollingTimeWindowMeanEnricher.java:
```diff
@@ -103,31 +119,69 @@ public void onEvent(SensorEvent<T> event) {
             return lastAverage = new ConfidenceQualifiedNumber(lastAverage.value, 0.0d);
         }
+        
+        // (previously there was an old comment here, pre-Jul-2014,  
```
Include a `TODO` in this comment - it's the sort of comment we want to delete if/when we figure out what the old comment meant, or we're happy with it. Hence it justifies a TODO because we want to change it.

ahgittin replied:
In that case I'll take it out

aledsage commented on policy/src/main/java/brooklyn/enricher/RollingTimeWindowMeanEnricher.java:
```diff
@@ -103,31 +119,69 @@ public void onEvent(SensorEvent<T> event) {
+        }
     }
     
     public ConfidenceQualifiedNumber getAverage() {
```
Nothing seems to call the `getAverage()` method, so the `TIMESTAMP_GRACE_TIME` is never used. Or am I missing something?

aledsage commented:
@ahgittin I'm not sure I follow the fix.

Is the main fix that `BasicSensorEvent` now defaults to the current timestamp, rather than 0?

I think I follow the change to `Transformer`, but what about `Propagator` where it calls `emitAllAttributes` when first setting all entities? Would the same logic as for Transformer not apply for Propagator as well?

ahgittin commented:
thanks @aledsage; i've implemented your suggestions, apart from `Propagator` where I've now added a TODO.

you are right in your thinking that we want this for `Propagator`; I simply hadn't noticed it.  it was `Transformer` that was causing the errors i was encountering.  however fixing it for propagator seems a little less clear-cut, and i've not tested it, so all i have done is the todo.

I think to fix, we need to make a timestamp a first-class property of sensors.

okay to merge as changed?

aledsage commented:
Looks good; merging.
